src/bsfs.c

```diff
@@ -1,34 +1,261 @@
+  pthread_rwlock_destroy(&file->lock);
+  free(file);
+}
+
+static int oft_realloc_buckets(bs_oft_t* table, size_t bucket_count) {
+  if (!bucket_count || bucket_count & (bucket_count - 1)) {
+    // not a power of 2
+    return -EINVAL;
+  }
+
+  bs_file_t** new_buckets =
+      (bs_file_t**) calloc(bucket_count, sizeof(bs_file_t*));
+  if (!new_buckets) {
+    return -ENOMEM;
+  }
+  free(table->buckets);
+  table->buckets = new_buckets;
+  table->bucket_count = bucket_count;
+  return 0;
+}
+
+static size_t oft_bucket_of(bft_offset_t index, size_t bucket_count) {
+  return (size_t) index & (bucket_count - 1); // assumes power-of-2 bucket count
+}
+
+static bool oft_matches_bucket(bs_file_t file, size_t bucket,
+                               size_t bucket_count) {
+  return file && oft_bucket_of(file->index, bucket_count) == bucket;
+}
+
+static bs_file_t oft_find(bs_oft_t* table, bft_offset_t index) {
+  size_t bucket = oft_bucket_of(index, table->bucket_count);
+  bs_file_t* prev_link = table->buckets[bucket];
+
+  if (prev_link) {
+    for (bs_file_t iter = *prev_link;
+         oft_matches_bucket(iter, bucket, table->bucket_count);
+         iter = iter->next) {
+      if (iter->index == index) {
+        return iter;
+      }
+    }
+  }
+
+  return NULL;
+}
+
+static void oft_do_insert(bs_oft_t* table, bs_file_t file) {
+  size_t bucket = oft_bucket_of(file->index, table->bucket_count);
+  if (table->buckets[bucket]) {
+    // insert after existing node
+    bs_file_t* prev_link = table->buckets[bucket];
+    file->next = *prev_link;
+    *prev_link = file;
+  } else {
+    // insert at head
+    file->next = table->head;
+    table->head = file;
+    table->buckets[bucket] = &table->head;
+
+    if (file->next) {
+      // patch other bucket with new next pointer
+      size_t next_bucket =
+          oft_bucket_of(file->next->index, table->bucket_count);
+      table->buckets[next_bucket] = &file->next;
+    }
+  }
+}
+
+static int oft_remove(bs_oft_t* table, bs_file_t file) {
+  size_t bucket = oft_bucket_of(file->index, table->bucket_count);
+  if (!table->buckets[bucket]) {
+    return -EINVAL;
+  }
+
+  bs_file_t* prev_link = table->buckets[bucket];
+  for (; *prev_link != file &&
+         oft_matches_bucket(*prev_link, bucket, table->bucket_count);
+       prev_link = &(*prev_link)->next) {
+  }
+
+  if (*prev_link != file) {
+    return -EINVAL;
+  }
+
+  if (*table->buckets[bucket] == file &&
+      !oft_matches_bucket(file->next, bucket, table->bucket_count)) {
+    // `file` was the only entry in the bucket - empty it
+    table->buckets[bucket] = NULL;
+  }
+
+  *prev_link = file->next;
+
+  if (*prev_link) {
+    // patch other bucket with new next pointer
+    size_t next_bucket =
+        oft_bucket_of((*prev_link)->index, table->bucket_count);
+    table->buckets[next_bucket] = prev_link;
+  }
+
+  table->size--;
+
+  return 0;
+}
+
+static int oft_rehash(bs_oft_t* table, size_t bucket_count) {
```

Comment:
`bucket_count` should be called here `new_bucket_count` (or something like that).